File: targets/java-web/src/main/java/com/example/vulnapp/vulns/Vuln16VerboseError.java
package com.example.vulnapp.vulns;

import com.example.vulnapp.infra.Db;
import org.springframework.web.bind.annotation.*;
import java.io.*;
import java.sql.*;

@RestController
public class Vuln16VerboseError {
    private final Db db;
    public Vuln16VerboseError(Db db) { this.db = db; }

    @GetMapping("/vuln16/lookup")
    public String lookup(@RequestParam String id) {
        try (Connection c = db.getConnection(); Statement st = c.createStatement()) {
            st.executeQuery("SELECT * FROM users WHERE id = " + id);
            return "{\"status\":\"ok\"}";
        } catch (Exception e) {
            // VULN:VULN-16:CWE-209:logic full exception / stack trace returned to the client
            StringWriter sw = new StringWriter();
            e.printStackTrace(new PrintWriter(sw));
            return "error: " + e + "\n" + sw;
        }
    }
}

File: targets/kotlin-web/src/main/kotlin/com/example/vulnapp/vulns/Vuln16VerboseError.kt
package com.example.vulnapp.vulns

import com.example.vulnapp.infra.Db
import org.springframework.web.bind.annotation.*


@RestController
class Vuln16VerboseError(private val db: Db) {

    @GetMapping("/vuln16/lookup")
    fun lookup(@RequestParam id: String): String = try {
        db.connection().use { c -> c.createStatement().use { st -> st.executeQuery("SELECT * FROM users WHERE id = $id") } }
        "{\"status\":\"ok\"}"
    } catch (e: Exception) {
        // VULN:VULN-16:CWE-209:logic full exception / stack trace returned to the client
        val sw = java.io.StringWriter(); e.printStackTrace(java.io.PrintWriter(sw)); "error: $e\n$sw"
    }
}
